File: android/src/main/java/io/mobilecron/NativeAgentBridge.kt
package io.mobilecron

import android.content.Context
import com.t6x.plugins.nativeagent.MemoryProviderImpl
import com.t6x.plugins.nativeagent.NativeNotifierImpl
import java.util.concurrent.atomic.AtomicBoolean
import uniffi.native_agent_ffi.NativeAgentHandle
import uniffi.native_agent_ffi.createHandleFromPersistedConfig

object NativeAgentBridge {
    private const val STORAGE_FILE = "CapacitorStorage"
    private const val CONFIG_PATH_KEY = "mobilecron:native-agent-config-path"

    @Volatile
    private var handle: NativeAgentHandle? = null
    private val libraryLoaded = AtomicBoolean(false)

    fun handleWake(context: Context, source: String): Boolean {
        return runCatching {
            val agentHandle = getOrCreateHandle(context) ?: return false
            agentHandle.handleWake(source)
            true
        }.getOrDefault(false)
    }

    private fun getOrCreateHandle(context: Context): NativeAgentHandle? {
        handle?.let { return it }
        synchronized(this) {
            handle?.let { return it }
            ensureLibraryLoaded()
            val configPath = context
                .getSharedPreferences(STORAGE_FILE, Context.MODE_PRIVATE)
                .getString(CONFIG_PATH_KEY, null)
                ?: return null

            val agentHandle = runCatching {
                createHandleFromPersistedConfig(configPath)
            }.getOrNull() ?: return null

            runCatching {
                agentHandle.setNotifier(NativeNotifierImpl(context.applicationContext))
            }
            runCatching {
                val memoryProvider = MemoryProviderImpl(context.applicationContext)
                if (memoryProvider.isAvailable()) {
                    agentHandle.setMemoryProvider(memoryProvider)
                }
            }

            handle = agentHandle
            return agentHandle
        }
    }

    private fun ensureLibraryLoaded() {
        if (libraryLoaded.get()) {
            return
        }
        synchronized(libraryLoaded) {
            if (libraryLoaded.get()) {
                return
            }
            System.loadLibrary("native_agent_ffi")
            libraryLoaded.set(true)
        }
    }
}

File: ios/Plugin/NativeAgentBridge.swift
import Foundation

#if canImport(CapacitorNativeAgent)
import CapacitorNativeAgent
#elseif canImport(NativeAgentPlugin)
import NativeAgentPlugin
#endif

#if canImport(CapacitorNativeAgent) || canImport(NativeAgentPlugin)
final class NativeAgentBridge {
    static let shared = NativeAgentBridge()

    private let configPathKey = "mobilecron:native-agent-config-path"
    private let lock = NSLock()
    private var handle: NativeAgentHandle?

    func handleWake(_ source: String) -> Bool {
        guard let handle = getOrCreateHandle() else {
            return false
        }
        do {
            try handle.handleWake(source: source)
            return true
        } catch {
            return false
        }
    }

    private func getOrCreateHandle() -> NativeAgentHandle? {
        lock.lock()
        defer { lock.unlock() }

        if let handle {
            return handle
        }

        guard let configPath = UserDefaults.standard.string(forKey: configPathKey) else {
            return nil
        }

        do {
            let handle = try createHandleFromPersistedConfig(configPath: configPath)
            try handle.setNotifier(notifier: NativeNotifierImpl())
            if let memoryProvider = MemoryProviderImpl.makeIfAvailable() {
                try handle.setMemoryProvider(provider: memoryProvider)
            }
            self.handle = handle
            return handle
        } catch {
            return nil
        }
    }
}
#else
final class NativeAgentBridge {
    static let shared = NativeAgentBridge()

    func handleWake(_ source: String) -> Bool {
        false
    }
}
#endif
